### PR TITLE
Cancel `WebSocketTestSession` on close

### LINKED ISSUE #947: Background Thread created by WebSocketTestSession should exit when close() is called
Hello!

I've been trying to write tests for a starlette websocket API. 

Only one problem - the server isn't written in a way to exit nicely when a websocket is closed. 
As a result, the tests just sit there, because `websocket_connect()` won't exit until the server does.

By looking at the source, I figured that `WebSocketTestSession` is waiting for the session thread (`_run()`) to complete.

Can we please make it so that it stops the server when `close()` is called?

---

I've cobbled together a small [commit](https://github.com/devxpy/starlette/commit/41907e863220b8b34f6bcc0d8f3a8d1241ded8eb) to make this work. If you're on board with this change, I'd love to upstream this.

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/starlette/issues/947">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/starlette/issues/947/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Kludex**
- Closes https://github.com/encode/starlette/issues/947

### CODE REVIEW

File: starlette/testclient.py [hunk 1]
```diff
@@ -113,14 +127,21 @@ async def _run(self) -> None:
...
+                ...
+            except BaseException as exc:
+                self._send_queue.put(exc)
+                raise
+            tg.cancel_scope.cancel()
+
+        async with anyio.create_task_group() as tg:
+            tg.start_soon(run_app, tg)
+            await self.event.wait()
+            tg.cancel_scope.cancel()
```
Comment (Kludex) [inline starlette/testclient.py]: This is the relevant code.

File: starlette/testclient.py [hunk 2]
```diff
@@ -99,10 +109,14 @@ def __enter__(self) -> "WebSocketTestSession":
...
         return self
 
+    async def _notify_close(self) -> None:
+        self.event.set()
+
     def __exit__(self, *args: typing.Any) -> None:
         try:
             self.close(1000)
         finally:
+            self.portal.start_task_soon(self._notify_close)
```
Comment (Kludex) [inline starlette/testclient.py:119]: And this.

File: tests/test_testclient.py [hunk 3]
```diff
@@ -254,9 +254,25 @@ async def asgi(receive, send):
...
+            await websocket.accept()
+            while True:
+                await anyio.sleep(0.1)
+
+        return asgi
+
+    client = test_client_factory(app)
+    with client.websocket_connect("/") as websocket:
+        ...
+    assert websocket.event.is_set()
```
Comment (Kludex) [inline tests/test_testclient.py]: This is the relevant test.

File: starlette/testclient.py [hunk 4]
```diff
@@ -1,16 +1,20 @@
+from __future__ import annotations
```
Comment (Kludex) [inline starlette/testclient.py:1]: If reviewer complains about this, I can undo the changes.

**Comment by Kludex** [comment]
@encode/maintainers someone? 🙏

File: starlette/testclient.py [hunk 5]
```diff
@@ -113,14 +127,21 @@ async def _run(self) -> None:
...
+
+        async def run_app(tg: anyio.abc.TaskGroup) -> None:
+            try:
+                await self.app(self.scope, self._asgi_receive, self._asgi_send)
+            except anyio.get_cancelled_exc_class():
+                ...
+            except BaseException as exc:
+                self._send_queue.put(exc)
+                raise
+            tg.cancel_scope.cancel()
```
Comment (frankie567) [inline starlette/testclient.py]: Shouldn't we also cancel the task group before raising the exception just above?

**Review by frankie567 — APPROVED** [review]
LGTM! This approach with the event flag makes total sense.

File: starlette/testclient.py (hunk 5 above)
Comment (Kludex) [reply starlette/testclient.py]: I did that now.

Comment (Kludex) [reply starlette/testclient.py]: Thanks.

**Pull request merged**